File: app/src/main/java/com/przemolab/oknotifier/modules/NotifierRepository.java
package com.przemolab.oknotifier.modules;

import android.content.Context;
import android.database.Cursor;
import android.net.Uri;

import com.przemolab.oknotifier.data.NotifierContract;
import com.przemolab.oknotifier.enums.SortOrder;
import com.przemolab.oknotifier.models.Contest;
import com.przemolab.oknotifier.models.Contestant;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import timber.log.Timber;

public class NotifierRepository {

    private Context context;

    public NotifierRepository(Context context) {
        this.context = context;
    }

    public List<Contest> getAll(SortOrder sortOrder) {
        try {
            List<Contest> contests = new ArrayList<>();
            Uri contestsUri = NotifierContract.ContestEntry.CONTENT_URI;
            Cursor cursor = context.getContentResolver()
                    .query(contestsUri, null, null, null, getSortOrder(sortOrder));

            if (cursor != null) {
                while (cursor.moveToNext()) {
                    contests.add(Contest.getFromCursor(cursor));
                }

                cursor.close();
            }

            return contests;
        } catch (Exception ex) {
            Timber.e(ex);
            return null;
        }
    }

    public List<Contest> getSubscribed() {
        try {
            List<Contest> subscribedContests = new ArrayList<>();
            Uri contestsUri = NotifierContract.ContestEntry.CONTENT_URI;
            Cursor cursor = context.getContentResolver()
                    .query(contestsUri, null, NotifierContract.ContestEntry.COLUMN_IS_SUBSCRIBED + "=1", null, null);

            if (cursor != null) {
                while (cursor.moveToNext()) {
                    subscribedContests.add(Contest.getFromCursor(cursor));
                }

                cursor.close();
            }

            return subscribedContests;
        } catch (Exception ex) {
            Timber.e(ex);
            return null;
        }
    }

    public void persistContests(List<Contest> contests) {
        try {
            List<Contest> persistedContests = getAll(SortOrder.SubscribedFirst);
            List<Contest> updatedContests = new ArrayList<>();

            for (Contest contest : contests) {
                boolean exists = false;

                for (Contest persistedContest : persistedContests) {
                    if (persistedContest.getContestId().equals(contest.getContestId())) {
                        contest.setSubscribed(persistedContest.isSubscribed());
                        exists = true;
                        break;
                    }
                }

                if (exists) {
                    updateContest(contest);
                    updatedContests.add(contest);
                } else {
                    createContest(contest);
                }
            }

            List<Contest> contestsToDelete = new ArrayList<>();
            for (Contest persistedContest: persistedContests) {
                boolean toDelete = true;

                for (Contest updatedContest: updatedContests) {
                    if (persistedContest.getContestId().equals(updatedContest.getContestId())) {
                        toDelete = false;
                        break;
                    }
                }

                if (toDelete) {
                    contestsToDelete.add(persistedContest);
                }
            }

            if (contestsToDelete.size() > 0) {
                deleteContests(contestsToDelete);
            }

        } catch (Exception ex) {
            Timber.e(ex);
        }
    }

    public List<Contestant> getAllContestants(String contestId) {
        try {
            List<Contestant> contestants = new ArrayList<>();
            Uri contestantsUri = NotifierContract.ContestantEntry.CONTENT_URI.buildUpon()
                    .appendPath("byContestId")
                    .appendPath(contestId)
                    .build();
            String sortOrder = NotifierContract.ContestantEntry.COLUMN_PROBLEMS_SOLVED + " DESC, " + NotifierContract.ContestantEntry.COLUMN_TIME + " DESC";
            Cursor cursor = context.getContentResolver()
                    .query(contestantsUri, null, null, null, sortOrder);

            if (cursor != null) {
                while (cursor.moveToNext()) {
                    contestants.add(Contestant.getFromCursor(cursor));
                }

                cursor.close();
            }

            return contestants;
        } catch (Exception ex) {
            Timber.e(ex);
            return null;
        }
    }

    public void persistContestants(String contestId, List<Contestant> contestants) {
        Timber.d("Creating contest");
        try {
            List<Contestant> persistedContestants = getAllContestants(contestId);

            for (Contestant contestant : contestants) {
                boolean exists = false;
                String contestantId = String.format("%s - %s", contestant.getName(), contestant.getContestId());

                for (Contestant persistedContestant : persistedContestants) {
                    String persistedContestantId = String.format("%s - %s", persistedContestant.getName(), persistedContestant.getContestId());
                    if (persistedContestantId.equals(contestantId)) {
                        contestant.setId(persistedContestant.getId());
                        exists = true;
                        break;
                    }
                }

                if (exists) {
                    updateContestant(contestant);
                } else {
                    createContestant(contestant);
                }
            }
        } catch (Exception ex) {
            Timber.e(ex);
        }
    }

    private void createContestant(Contestant contestant) {
        Uri uri = NotifierContract.ContestantEntry.CONTENT_URI;

        Timber.d("Creating contestant: %s [%s]", contestant.getName(), contestant.getContestId());
        context.getContentResolver().insert(uri, contestant.toContentValues());
    }

    private void updateContestant(Contestant contestant) {
        Uri uri = NotifierContract.ContestantEntry.CONTENT_URI;

        Timber.d("Updating contest: %s [%s]", contestant.getName(), contestant.getContestId());
        context.getContentResolver().update(uri, contestant.toContentValues(),
                NotifierContract.ContestantEntry._ID + "=?",
                new String[] {String.valueOf(contestant.getId())});
    }

    private void createContest(Contest contest) {
        Uri uri = NotifierContract.ContestEntry.CONTENT_URI;

        Timber.d("Creating contest: %s [%s]", contest.getName(), contest.getContestId());
        context.getContentResolver().insert(uri, contest.toContentValues());
    }

    public void updateContest(Contest contest) {
        Uri uri = NotifierContract.ContestEntry.CONTENT_URI;

        Timber.d("Updating contest: %s [%s]", contest.getName(), contest.getContestId());
        context.getContentResolver().update(uri, contest.toContentValues(),
                NotifierContract.ContestEntry.COLUMN_CONTEST_ID + "=?",
                new String[] {contest.getContestId()});
    }

    private void deleteContests(List<Contest> contestsToDelete) {
        Uri deleteUri = NotifierContract.ContestEntry.CONTENT_URI
                .buildUpon().appendPath("byContestIds").build();

        String[] contestsIds = new String[contestsToDelete.size()];
        for (int i = 0; i < contestsToDelete.size(); i++) {
            contestsIds[i] = contestsToDelete.get(i).getContestId();
        }

        Timber.d("Deleting contests: %s", Arrays.toString(contestsIds));
        context.getContentResolver().delete(deleteUri, null, contestsIds);
    }

    private String getSortOrder(SortOrder sortOrder) {
        String orderBy = NotifierContract.ContestEntry.COLUMN_IS_SUBSCRIBED + " DESC";
        switch (sortOrder) {
            case ByName:
                orderBy = NotifierContract.ContestEntry.COLUMN_NAME;
                break;
            case ByStartDate:
                orderBy = NotifierContract.ContestEntry.COLUMN_START_DATE;
                break;
            case ByNumberOfProblems:
                orderBy = NotifierContract.ContestEntry.COLUMN_NUM_OF_PROBLEMS + " DESC";
                break;
            case ByNumberOfContestants:
                orderBy = NotifierContract.ContestEntry.COLUMN_NUM_OF_CONTESTANTS + " DESC";
                break;
        }

        return orderBy;
    }
}

File: app/src/main/java/com/przemolab/oknotifier/modules/NotifierRepository.kt
package com.przemolab.oknotifier.modules

import android.content.Context
import com.przemolab.oknotifier.AppExecutors
import com.przemolab.oknotifier.data.AppDatabase
import com.przemolab.oknotifier.data.entries.ContestEntry

import com.przemolab.oknotifier.data.entries.ContestantEntry
import com.przemolab.oknotifier.enums.SortOrder
import com.przemolab.oknotifier.interfaces.INotifierRepository

import java.util.ArrayList

import timber.log.Timber

class NotifierRepository(context: Context) : INotifierRepository {

    val db = AppDatabase.getInstance(context)!!

    override val subscribed: List<ContestEntry>?
        get() {
            return try {
                db.contestDao().getSubscribed()
            } catch (ex: Exception) {
                Timber.e(ex)
                null
            }
        }

    override fun getAllContests(sortOrder: SortOrder): List<ContestEntry>? {
        return try {
            when (sortOrder) {
                SortOrder.SubscribedFirst -> db.contestDao().getAll().sortedByDescending { it -> it.subscribed }
                SortOrder.ByNumberOfProblems -> db.contestDao().getAll().sortedByDescending { it -> it.numberOfProblems }
                SortOrder.ByNumberOfContestants -> db.contestDao().getAll().sortedByDescending { it -> it.numberOfContestants }
                SortOrder.ByStartDate -> db.contestDao().getAll().sortedBy { it -> it.startDate }
                else ->
                    db.contestDao().getAll().sortedBy { it -> it.name }
            }
        } catch (ex: Exception) {
            Timber.e(ex)
            null
        }
    }

    override fun persistContests(contestEntries: List<ContestEntry>?) {
        try {
            val persistedContests = getAllContests(SortOrder.SubscribedFirst)
            val updatedContests = ArrayList<ContestEntry>()

            for (contest in contestEntries!!) {
                var exists = false

                for (persistedContest in persistedContests!!) {
                    if (persistedContest.contestId == contest.contestId) {
                        contest.subscribed = persistedContest.subscribed
                        exists = true
                        break
                    }
                }

                if (exists) {
                    updateContest(contest)
                    updatedContests.add(contest)
                } else {
                    createContest(contest)
                }
            }

            val contestsToDelete = ArrayList<ContestEntry>()
            for (persistedContest in persistedContests!!) {
                var toDelete = true

                for (updatedContest in updatedContests) {
                    if (persistedContest.contestId == updatedContest.contestId) {
                        toDelete = false
                        break
                    }
                }

                if (toDelete) {
                    contestsToDelete.add(persistedContest)
                }
            }

            if (contestsToDelete.size > 0) {
                deleteContests(contestsToDelete)
            }

        } catch (ex: Exception) {
            Timber.e(ex)
        }
    }

    override fun updateContest(contestEntry: ContestEntry) {
        Timber.d("Updating contest: %s [%s]", contestEntry.name, contestEntry.contestId)
        AppExecutors.getInstance()?.diskIO?.execute {
            kotlin.run {
                db.contestDao().update(contestEntry)
            }
        }
    }

    override fun getAllContestants(contestId: String): List<ContestantEntry>? {
        return try {
            db.contestantDao().getByContestId(contestId)
        } catch (ex: Exception) {
            Timber.e(ex)
            null
        }
    }

    override fun persistContestants(contestId: String, contestants: List<ContestantEntry>?) {
        Timber.d("Persisting contestants")
        try {
            val persistedContestants = getAllContestants(contestId)!!

            for (contestant in contestants!!) {
                var exists = false
                val contestantId = String.format("%s - %s", contestant.name, contestant.contestId)

                for (persistedContestant in persistedContestants) {
                    val persistedContestantId = String.format("%s - %s", persistedContestant.name, persistedContestant.contestId)
                    if (persistedContestantId == contestantId) {
                        contestant.id = persistedContestant.id
                        exists = true
                        break
                    }
                }

                if (exists) {
                    updateContestant(contestant)
                } else {
                    createContestant(contestant)
                }
            }

            if (contestants.size > persistedContestants.size) {
                updateNumberOfContestants(contestId, contestants.size)
            }
        } catch (ex: Exception) {
            Timber.e(ex)
        }
    }

    private fun updateNumberOfContestants(contestId: String, numberOfContestants: Int) {
        Timber.d("Updating # of contestants in contest: %s [%s]", contestId, numberOfContestants)
        AppExecutors.getInstance()?.diskIO?.execute {
            kotlin.run {
                db.contestDao().updateNumberOfContestants(contestId, numberOfContestants)
            }
        }
    }

    private fun createContestant(contestantEntry: ContestantEntry) {
        Timber.d("Creating contestant: %s [%s]", contestantEntry.name, contestantEntry.contestId)
        AppExecutors.getInstance()?.diskIO?.execute {
            kotlin.run {
                db.contestantDao().insert(contestantEntry)
            }
        }
    }

    private fun updateContestant(contestantEntry: ContestantEntry) {
        Timber.d("Updating contestant: %s [%s]", contestantEntry.name, contestantEntry.contestId)
        AppExecutors.getInstance()?.diskIO?.execute {
            kotlin.run {
                db.contestantDao().update(contestantEntry)
            }
        }
    }

    private fun createContest(contestEntry: ContestEntry) {
        Timber.d("Creating contest: %s [%s]", contestEntry.name, contestEntry.contestId)
        AppExecutors.getInstance()?.diskIO?.execute {
            kotlin.run {
                db.contestDao().insert(contestEntry)
            }
        }
    }

    private fun deleteContests(contestsToDelete: List<ContestEntry>) {
        Timber.d("Deleting contests: %s", contestsToDelete.map { it -> it.contestId })
        AppExecutors.getInstance()?.diskIO?.execute {
            kotlin.run {
                db.contestDao().deleteMany(contestsToDelete)
            }
        }
    }
}
